In [37]:
import re
import sys
import pandas as pd
import numpy as np
from github import Github
from getpass import getpass
from pandas.io.json import json_normalize
from IPython.display import display, HTML
import matplotlib.pyplot as plt
import networkx as nx
%matplotlib inline

In [6]:
repos = pd.read_csv("awe_repos_100_in_cats.csv", encoding="utf-8")

In [42]:
ACCESS_TOKEN = getpass(prompt="Enter github token:")
client = Github(ACCESS_TOKEN, per_page=100)

Enter github token:········



In [49]:
g = nx.DiGraph() #or read it from disk #g = nx.read_gpickle("resources/ch07-github/data/github.gpickle.3")
cache = {
    'repos':{},
    'contributors':{},
    'followers':{}
}

In [ ]:
%%time

START_FROM = 103 #index
for i, repo_name in enumerate(repos.iloc[START_FROM:]["repo"].values):
    if ((START_FROM+i) % 100) == 0: print(START_FROM+i,repo_name) 
    try:
        repo = client.get_repo(repo_name)
        g.add_node(repo.name + '(repo)', 
                   type='repo', 
                   lang=repo.language, 
                   owner=repo.owner.login,
                   description=repo.description
                  )
        cache['repos'][repo_name] = repo
        
        for contributor in repo.get_contributors():
            g.add_node(contributor.login + '(user)', type='user')
            g.add_edge(contributor.login + '(user)', repo.name + '(repo)', type='contributors')
            cache['contributors'][contributor.login] = contributor
            for follower in contributor.get_followers():
                if follower.login + '(user)' in g:
                    g.add_edge(follower.login + '(user)', contributor.login + '(user)', 
                           type='follows')
                    cache['followers'][follower.login] = follower
        
    except Exception, e: #ssl.SSLError
        print >> sys.stderr, "Encountered an error fetching data for", \
                             repo_name, "Skipping."
        print >> sys.stderr, e

    #if ((START_FROM+i) % 100) == 0: print "Processed", START_FROM+i+1, " ", repo_name," nodes. Num nodes/edges in graph", \
    #      g.number_of_nodes(), "/", g.number_of_edges()
    if client.rate_limiting < 400: print("Rate limit remaining", client.rate_limiting)
    
nx.write_gpickle(g, "github.LONG.gpickle")    

Encountered an error fetching data for tj/node-progress Skipping.
unsupported operand type(s) for +: 'NoneType' and 'str'
Encountered an error fetching data for bucaran/fly Skipping.
unsupported operand type(s) for +: 'NoneType' and 'str'
Encountered an error fetching data for voodootikigod/node-serialport Skipping.
unsupported operand type(s) for +: 'NoneType' and 'str'


(200, u'mafintosh/peek-stream')


Encountered an error fetching data for substack/stream-combiner2 Skipping.
('_ssl.c:565: The handshake operation timed out',)
Encountered an error fetching data for nodejs/readable-stream Skipping.
timed out
Encountered an error fetching data for almost/through2-concurrent Skipping.
timed out
Encountered an error fetching data for sockjs/sockjs-node Skipping.
timed out
Encountered an error fetching data for SocketCluster/socketcluster Skipping.
timed out
Encountered an error fetching data for primus/primus Skipping.
timed out
Encountered an error fetching data for simonswain/straw Skipping.
timed out
Encountered an error fetching data for fed135/Kalm Skipping.
timed out
Encountered an error fetching data for lovell/sharp Skipping.
timed out
Encountered an error fetching data for sindresorhus/image-type Skipping.
timed out
Encountered an error fetching data for aheckmann/gm Skipping.
timed out
Encountered an error fetching data for EyalAr/lwip Skipping.
timed out
Encountered an error fe

In [58]:
len(cache["repos"])

102

In [59]:
nx.info(g)

'Name: \nType: DiGraph\nNumber of nodes: 1770\nNumber of edges: 8115\nAverage in degree:   4.5847\nAverage out degree:   4.5847'

In [60]:
nx.write_gpickle(g, "github.LONG.gpickle")

In [62]:
mtsw_users = [n for n in g if g.node[n]['type'] == 'user']
h = g.subgraph(mtsw_users)

#nx.write_gexf(h, "github2.gexf")
#nx.write_gml(h, "github2.gml")
nx.write_graphml(h, "github1770.graphml")

Encountered an error fetching data for substack/stream-combiner2 Skipping.
('_ssl.c:565: The handshake operation timed out',)
Encountered an error fetching data for nodejs/readable-stream Skipping.
timed out
Encountered an error fetching data for almost/through2-concurrent Skipping.
timed out
Encountered an error fetching data for sockjs/sockjs-node Skipping.
timed out
Encountered an error fetching data for SocketCluster/socketcluster Skipping.
timed out
Encountered an error fetching data for primus/primus Skipping.
timed out
Encountered an error fetching data for simonswain/straw Skipping.
timed out
Encountered an error fetching data for fed135/Kalm Skipping.
timed out
Encountered an error fetching data for lovell/sharp Skipping.
timed out
Encountered an error fetching data for sindresorhus/image-type Skipping.
timed out
Encountered an error fetching data for aheckmann/gm Skipping.
timed out
Encountered an error fetching data for EyalAr/lwip Skipping.
timed out
Encountered an error fetching data for nodeca/pica Skipping.
timed out
Encountered an error fetching data for sindresorhus/is-progressive Skipping.
timed out
Encountered an error fetching data for nodeca/probe-image-size Skipping.
timed out
Encountered an error fetching data for epeli/underscore.string Skipping.
timed out
Encountered an error fetching data for ashtuchkin/iconv-lite Skipping.
timed out
Encountered an error fetching data for sindresorhus/repeating Skipping.
timed out
Encountered an error fetching data for sindresorhus/string-length Skipping.
timed out
Encountered an error fetching data for sindresorhus/camelcase Skipping.
timed out
Encountered an error fetching data for sindresorhus/escape-string-regexp Skipping.
timed out
Encountered an error fetching data for sindresorhus/execall Skipping.
timed out
Encountered an error fetching data for sindresorhus/splice-string Skipping.
timed out
Encountered an error fetching data for sindresorhus/indent-string Skipping.
timed out
Encountered an error fetching data for sindresorhus/strip-indent Skipping.
timed out
Encountered an error fetching data for sindresorhus/detect-indent Skipping.
timed out
Encountered an error fetching data for mathiasbynens/he Skipping.
timed out
Encountered an error fetching data for mashpie/i18n-node Skipping.
timed out
Encountered an error fetching data for nodeca/babelfish Skipping.
timed out
Encountered an error fetching data for sindresorhus/parse-columns Skipping.
timed out
Encountered an error fetching data for codekirei/hanging-indent Skipping.
timed out
Encountered an error fetching data for sindresorhus/matcher Skipping.
timed out
Encountered an error fetching data for sindresorhus/random-int Skipping.
timed out
Encountered an error fetching data for sindresorhus/random-float Skipping.
timed out
Encountered an error fetching data for sindresorhus/unique-random Skipping.
timed out
Encountered an error fetching data for sindresorhus/round-to Skipping.
timed out
Encountered an error fetching data for scijs/ndarray Skipping.
timed out
Encountered an error fetching data for josdejong/mathjs Skipping.
timed out
Encountered an error fetching data for sindresorhus/math-sum Skipping.
timed out
Encountered an error fetching data for sindresorhus/math-clamp Skipping.
timed out
Encountered an error fetching data for fibo/algebra Skipping.
timed out
Encountered an error fetching data for felixge/node-dateformat Skipping.
timed out
Encountered an error fetching data for samverschueren/tz-format Skipping.
timed out
Encountered an error fetching data for sindresorhus/normalize-url Skipping.
timed out
Encountered an error fetching data for sindresorhus/humanize-url Skipping.
timed out
Encountered an error fetching data for nodeca/url-unshort Skipping.
timed out
Encountered an error fetching data for pid/speakingurl Skipping.
timed out
Encountered an error fetching data for snd/url-pattern Skipping.
timed out
In [58]:
Encountered an error fetching data for wooorm/remark Skipping.
('The read operation timed out',)
Encountered an error fetching data for inikulin/parse5 Skipping.
timed out
Encountered an error fetching data for sindresorhus/strip-json-comments Skipping.
timed out
Encountered an error fetching data for sindresorhus/strip-css-comments Skipping.
timed out
Encountered an error fetching data for sindresorhus/parse-json Skipping.
timed out
Encountered an error fetching data for medialize/URI.js Skipping.
timed out
Encountered an error fetching data for postcss/postcss Skipping.
timed out
Encountered an error fetching data for dominictarr/JSONStream Skipping.
timed out
Encountered an error fetching data for mafintosh/csv-parser Skipping.
timed out
Encountered an error fetching data for sindresorhus/neat-csv Skipping.
timed out
Encountered an error fetching data for pegjs/pegjs Skipping.
timed out
Encountered an error fetching data for lapwinglabs/x-ray Skipping.
timed out
Encountered an error fetching data for Hardmath123/nearley Skipping.
timed out
Encountered an error fetching data for juliangruber/binary-extract Skipping.
timed out
Encountered an error fetching data for nemtsov/json-mask Skipping.
timed out
Encountered an error fetching data for stylecow/stylecow Skipping.
timed out
Encountered an error fetching data for nodeca/js-yaml Skipping.
timed out
Encountered an error fetching data for dominictarr/excel-stream Skipping.
timed out
Encountered an error fetching data for seegno/google-libphonenumber Skipping.
timed out
Encountered an error fetching data for sindresorhus/pretty-bytes Skipping.
timed out
Encountered an error fetching data for sindresorhus/pretty-ms Skipping.
timed out
Encountered an error fetching data for rauchg/ms.js Skipping.
timed out
Encountered an error fetching data for AriaMinaei/pretty-error Skipping.
timed out
Encountered an error fetching data for taijinlee/humanize Skipping.
timed out
Encountered an error fetching data for archiverjs/node-archiver Skipping.
('_ssl.c:565: The handshake operation timed out',)
Encountered an error fetching data for bower/decompress-zip Skipping.
timed out
Encountered an error fetching data for nodeca/pako Skipping.
timed out
Encountered an error fetching data for mafintosh/tar-stream Skipping.
timed out
Encountered an error fetching data for mafintosh/tar-fs Skipping.
timed out
Encountered an error fetching data for kevva/decompress Skipping.
timed out